### PR TITLE
Chopper and DeWulf Harvester

### PR DESCRIPTION
**Pull request by pvaiko**
In general, making harvester work which:
- always need an unloader, have no tank.
- have an autoaim pipe, so the unloader can move around more freely.

### CODE REVIEW

File: scripts/Course.lua [hunk 1]
```diff
@@ -1130,194 +1158,176 @@ end
...
-			DebugUtil.drawDebugLine(x, y + 3, z, nx, ny + 3, nz, 0, 0, 100)
-		end
-	end
-end
-
---- Waypoints generated by AD have no explicit reverse attribute, they infer it from elsewhere, see
---- https://github.com/Courseplay/courseplay/issues/7026#issuecomment-808715976
---- We assume all AD courses start forward and assume there's a direction change whenever there's an angle
---- over 100 degrees between two subsequent waypoints
-function Course:addReverseForAutoDriveCourse()
```
Comment (schwiti6190) [inline scripts/Course.lua:1285]: I need this function later down the road :)

File: scripts/ai/AIDriveStrategyUnloadCombine.lua [hunk 2]
```diff
@@ -592,8 +989,19 @@ function AIDriveStrategyUnloadCombine:onLastWaypointPassed()
     elseif self.state == self.states.BACKING_UP_FOR_REVERSING_COMBINE then
         self:setNewState(self.stateAfterMovedOutOfWay)
         self:startRememberedCourse()
+    elseif self.state == self.states.MOVING_BACK_BEFORE_PATHFINDING then
+        if self.state.properties.pathfinderController then
+            self:debug('Retry last pathfinding after moved back a bit')
+            self:setNewState(self.states.WAITING_FOR_PATHFINDER)
+            self.lastState.properties.pathfinderController:retry(self.lastState.properties.pathfinderContext)
```
Comment (schwiti6190) [inline scripts/ai/AIDriveStrategyUnloadCombine.lua:996]: I like the idea of this. If we want to use this more often, than a more generic idea would be great :)

File: scripts/ai/PathfinderController.lua [hunk 3]
```diff
@@ -305,6 +329,16 @@ function PathfinderController:findPathToGoal(context, goal, numRetries)
     return true
 end
 
+function PathfinderController:findAnalyticPathFromVehicleToGoal(goal, allowReverse)
```
Comment (schwiti6190) [inline scripts/ai/PathfinderController.lua:332]: Could you add a description what this function does?

File: scripts/ai/controllers/CombineController.lua [hunk 4]
```diff
@@ -156,14 +178,15 @@ function CombineController:fixDischargeDistanceForChopper()
...
         local safeDischargeNodeMaxDistance = 40
-        if dischargeNode.maxDistance < safeDischargeNodeMaxDistance then
+        if false and dischargeNode.maxDistance < safeDischargeNodeMaxDistance then
             self:debug('Chopper maximum throw distance is %.1f, increasing to %.1f', dischargeNode.maxDistance, safeDischargeNodeMaxDistance)
             dischargeNode.maxDistance = safeDischargeNodeMaxDistance
         end
     end
 end
 
 function CombineController:isChopper()
```
Comment (schwiti6190) [inline scripts/ai/controllers/CombineController.lua:188]: Do we want to remove this then ? Or is there any use in trying to separate a choper with the dewulf and so on?

**Review by schwiti6190 — CHANGES_REQUESTED** [review]
Minior changes :)

File: scripts/ai/PathfinderController.lua (hunk 3 above)
Comment (pvaiko) [reply scripts/ai/PathfinderController.lua:336]: Added

File: scripts/ai/AIDriveStrategyCombineCourse.lua [hunk 5]
```diff
@@ -1446,7 +1502,7 @@ end
 
 -- TODO: move this to the PipeController?
```
[the remark this answers is not in the document]
Comment (pvaiko) [reply scripts/ai/AIDriveStrategyCombineCourse.lua]: Removed

File: scripts/Course.lua (hunk 1 above)
Comment (pvaiko) [reply scripts/Course.lua:1285]: Added back in :)

File: scripts/ai/PathfinderController.lua (hunk 3 above)
Comment (pvaiko) [reply scripts/ai/PathfinderController.lua:336]: Added

File: scripts/ai/controllers/CombineController.lua (hunk 4 above)
Comment (pvaiko) [reply scripts/ai/controllers/CombineController.lua:187]: Removed call but left the function in, we'll see later if we need it.

**Pull request merged**